src/ray/gcs/client.cc

```diff
@@ -2,44 +2,135 @@
+  std::vector<std::string> addresses;
+  std::vector<int> ports;
+  GetRedisShards(primary_context_->sync_context(), &addresses, &ports);
+  if (addresses.size() == 0 || ports.size() == 0) {
+    addresses.push_back(address);
+    ports.push_back(port);
+  }
+
+  // First time connect: populate shard_contexts.
+  if (shard_contexts_.empty()) {
+    for (unsigned int i = 0; i < addresses.size(); ++i) {
+      // Slower than emplace but resource safe.
+      shard_contexts_.push_back(std::make_shared<RedisContext>());
+    }
+  }
+
+  // Call connect for all contexts. Safe to do many times.
+  // Here shard_contexts_.size() == addresses.size();
+  for (unsigned int i = 0; i < addresses.size(); ++i) {
+    RAY_CHECK_OK(shard_contexts_[i]->Connect(addresses[i], ports[i], DUMMY_BOOL));
+  }
+
+  client_table_.reset(new ClientTable({primary_context_}, this, client_id));
+  error_table_.reset(new ErrorTable({primary_context_}, this));
+  // Tables below would be sharded.
+  object_table_.reset(new ObjectTable(shard_contexts_, this, command_type));
+  actor_table_.reset(new ActorTable(shard_contexts_, this));
+  task_table_.reset(new TaskTable(shard_contexts_, this, command_type));
+  raylet_task_table_.reset(new raylet::TaskTable(shard_contexts_, this, command_type));
+  task_reconstruction_log_.reset(new TaskReconstructionLog(shard_contexts_, this));
+  task_lease_table_.reset(new TaskLeaseTable(shard_contexts_, this));
+  heartbeat_table_.reset(new HeartbeatTable(shard_contexts_, this));
+  profile_table_.reset(new ProfileTable(shard_contexts_, this));
   command_type_ = command_type;
 }
 
 #if RAY_USE_NEW_GCS
 // Use of kChain currently only applies to Table::Add which affects only the
 // task table, and when RAY_USE_NEW_GCS is set at compile time.
-AsyncGcsClient::AsyncGcsClient(const ClientID &client_id)
-    : AsyncGcsClient(client_id, CommandType::kChain) {}
+AsyncGcsClient::AsyncGcsClient(const std::string &address, int port, const ClientID &client_id)
+    : AsyncGcsClient(address, port, client_id, CommandType::kChain) {}
 #else
-AsyncGcsClient::AsyncGcsClient(const ClientID &client_id)
-    : AsyncGcsClient(client_id, CommandType::kRegular) {}
+AsyncGcsClient::AsyncGcsClient(const std::string &address, int port, const ClientID &client_id)
+    : AsyncGcsClient(address, port, client_id, CommandType::kRegular) {}
 #endif  // RAY_USE_NEW_GCS
 
-AsyncGcsClient::AsyncGcsClient(CommandType command_type)
-    : AsyncGcsClient(ClientID::from_random(), command_type) {}
+AsyncGcsClient::AsyncGcsClient(const std::string &address, int port, CommandType command_type)
+    : AsyncGcsClient(address, port, ClientID::from_random(), command_type) {}
 
-AsyncGcsClient::AsyncGcsClient() : AsyncGcsClient(ClientID::from_random()) {}
+AsyncGcsClient::AsyncGcsClient(const std::string &address, int port) : AsyncGcsClient(address, port, ClientID::from_random()) {}
 
 Status AsyncGcsClient::Connect(const std::string &address, int port, bool sharding) {
```

Comment:
Is it possible to remove this method now?

Reply:
There is a //TODO here which may need us to call client table's connect. I moved it up to the ctor